V8 v8.9 release blog post (#523)
v8/v8.dev: src/blog/v8-release-89.md

---
title: 'V8 release v8.9'
author: 'Ingvar Stepanyan, awaiting a call'
avatars:
 - 'ingvar-stepanyan'
date: 2021-02-03
tags:
 - release
description: 'V8 release v8.9 brings performance improvements to calls with argument size mismatch.'
tweet: ''
---
Every six weeks, we create a new branch of V8 as part of our [release process](https://v8.dev/docs/release-process). Each version is branched from V8’s Git master immediately before a Chrome Beta milestone. Today we’re pleased to announce our newest branch, [V8 version 8.9](https://chromium.googlesource.com/v8/v8.git/+log/branch-heads/8.9), which is in beta until its release in coordination with Chrome 89 Stable in several weeks. V8 v8.9 is filled with all sorts of developer-facing goodies. This post provides a preview of some of the highlights in anticipation of the release.

## JavaScript

> ### Top-level await

Reviewer: ```suggestion
### Top-level `await`
```

> [Top-level await](https://github.com/tc39/proposal-top-level-await) is available in the [Blink rendering engine](https://www.chromium.org/blink) 89, a primary embedder of V8.

Reviewer: ```suggestion
[Top-level `await`](https://v8.dev/features/top-level-await) is available in the [Blink rendering engine](https://www.chromium.org/blink) 89, a primary embedder of V8.
```

Reviewer: Best to consistently point to our feature explainer IMHO. People can find the proposal link from there if they want more info.

Author: I thought about this too, but wasn't sure because it kind of renders the following sentence useless? (its only purpose is to provide a link to our explainer)

Author: Should I remove the follow-up sentence?

Reviewer: Up to you :) I think it's fine either way

> In standalone V8, top-level await remains behind the `--harmony-top-level-await` flag.

Reviewer: ```suggestion
In standalone V8, top-level `await` remains behind the `--harmony-top-level-await` flag.
```

Please see [our explainer](https://v8.dev/features/top-level-await) for more details.

## Performance

### Faster calls with arguments size mismatch

JavaScript allows calling a function with a different number of arguments than the expected number of parameters, i.e., one can pass either fewer or more arguments than the declared formal parameters. The former case is called under-application and the latter is called over-application.

In the under-application case, the remaining parameters get assigned to the `undefined` value. In the over-application case, the remaining arguments can be either accessed by using the rest parameter and the `Function.prototype.arguments` property, or they are simply superfluous and ignored. Many web and Node.js frameworks nowadays use this JS feature to accept optional parameters and create a more flexible API.

> Until recently, V8 had a special machinery to deal with arguments size mismatch: the arguments adaptor frame. Unfortunately, argument adaption comes at a performance cost and is commonly needed in modern front-end and middleware frameworks. It turns out that with a clever design (like reversing the order of the arguments in the stack), we can remove this extra frame, simplify V8 codebase and get rid of almost the entire overhead.

Reviewer: ```suggestion
Until recently, V8 had a special machinery to deal with arguments size mismatch: the arguments adaptor frame. Unfortunately, argument adaption comes at a performance cost and is commonly needed in modern front-end and middleware frameworks. It turns out that with a clever design (like reversing the order of the arguments in the stack), we can remove this extra frame, simplify the V8 codebase, and get rid of the overhead almost entirely.
```
